code/espurna/sensors/EZOPHSensor.h

```diff
@@ -0,0 +1,142 @@
+
+        EZOPHSensor(): BaseSensor() {
+            _count = 1;
+            _sensor_id = SENSOR_EZOPH_ID;
+        }
+
+        ~EZOPHSensor() {
+            if (_serial) delete _serial;
+        }
+
+        // ---------------------------------------------------------------------
+
+        void setRX(unsigned char pin_rx) {
+            if (_pin_rx == pin_rx) return;
+            _pin_rx = pin_rx;
+            _dirty = true;
+        }
+
+        void setTX(unsigned char pin_tx) {
+            if (_pin_tx == pin_tx) return;
+            _pin_tx = pin_tx;
+            _dirty = true;
+        }
+
+        // ---------------------------------------------------------------------
+
+        unsigned char getRX() {
+            return _pin_rx;
+        }
+
+        unsigned char getTX() {
+            return _pin_tx;
+        }
+
+        // ---------------------------------------------------------------------
+        // Sensor API
+        // ---------------------------------------------------------------------
+
+        // Initialization method, must be idempotent
+        void begin() {
+            if (!_dirty) return;
+
+            if (_serial) delete _serial;
+
+            _serial = new SoftwareSerial(_pin_rx, _pin_tx);
+            _serial->begin(9600);
+
+            _ready = true;
+            _dirty = false;
+        }
+
+        // Descriptive name of the sensor
+        String description() {
+            char buffer[28];
+            snprintf(buffer, sizeof(buffer), "EZOPH @ SwSerial(%u,%u)", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Descriptive name of the slot # index
+        String slot(unsigned char index) {
+            return description();
+        };
+
+        // Address of the sensor (it could be the GPIO or I2C address)
+        String address(unsigned char index) {
+            char buffer[6];
+            snprintf(buffer, sizeof(buffer), "%u:%u", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Type for slot # index
+        unsigned char type(unsigned char index) {
+            if (index == 0) return MAGNITUDE_PH;
+            return MAGNITUDE_NONE;
+        }
+
+        void pre() {
+            _read();
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {
+            if (index == 0) return _ph;
+            return 0;
+        }
+
+
+
+    protected:
+
+        // ---------------------------------------------------------------------
+        // Protected
+        // ---------------------------------------------------------------------
+
+        void _read() {
+            String ph_serial = "";
+            ph_serial.reserve(30);
+
+            while ((_serial->available() > 0)) {
+                char ph_char = (char)_serial->read();
+                ph_serial += ph_char;
+
+                if (ph_char == '\r') {
+                  break;
+                }
+
+                yield();
```

Comment:
Is this considered best practice?

Reply:
Maybe not. Real serial [`.available()`](https://github.com/esp8266/Arduino/blob/42c977bd4d14281e46ff5d01f8f860dd0c195a08/cores/esp8266/HardwareSerial.cpp#L90-L97) already yields, SW one see link above.

Note: [`optimistic_yield()`](https://github.com/esp8266/Arduino/blob/42c977bd4d14281e46ff5d01f8f860dd0c195a08/cores/esp8266/core_esp8266_main.cpp#L109-L115) checks if loop() `now() - loop_start_timestamp` is bigger than given usec interval

Reply:
Exactly, and apparently the actual rxRead method already has `WAIT` which is a micro for 
`optimistic_yield()`.

Thanks for the comments, it has been a great learning exercise for me.

Reply:
Ok, removed.